scanner/services/indexer.go

```diff
@@ -0,0 +1,140 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type indexerService struct {
+	v4.UnimplementedIndexerServer
+	indexer indexer.Indexer
+}
+
+// NewIndexerService creates a new indexer service.
+func NewIndexerService(indexer indexer.Indexer) *indexerService {
+	return &indexerService{
+		indexer: indexer,
+	}
+}
+
+func (s *indexerService) CreateIndexReport(ctx context.Context, req *v4.CreateIndexReportRequest) (*v4.IndexReport, error) {
+	// TODO We currently only support container images, hence we assume the resource
+	//      is of that type. When introducing nodes and other resources, this should
+	//      evolve.
+	resourceType := "containerimage"
+	if err := validateContainerImageRequest(req); err != nil {
+		return nil, err
+	}
+	// Create a claircore.Digest using the full Hash ID, including resource type.
+	hashIDSum := sha512.Sum512([]byte(req.GetHashId()))
+	manifestDigest, err := claircore.NewDigest(claircore.SHA512, hashIDSum[:])
+	if err != nil {
+		return nil, fmt.Errorf("internal error: creating container image manifest digest: %w", err)
+	}
+
+	ctx = zlog.ContextWithValues(ctx,
+		"component", "scanner/service/indexer",
+		"resource_type", resourceType,
+		"manifest_digest", manifestDigest.String())
+
+	// Setup authentication.
+	var opts []indexer.Option
+	hasAuth := req.GetContainerImage().GetUsername() != ""
+	if hasAuth {
+		opts = append(opts, indexer.WithAuth(&authn.Basic{
+			Username: req.GetContainerImage().GetUsername(),
+			Password: req.GetContainerImage().GetPassword(),
+		}))
+	}
+
+	// Create index report.
+	zlog.Info(ctx).
+		Str("image_url", req.GetContainerImage().GetUrl()).
+		Bool("has_auth", hasAuth).
+		Msg("creating index report for container image")
+	clairReport, err := s.indexer.IndexContainerImage(
+		ctx,
+		manifestDigest,
+		req.GetContainerImage().GetUrl(),
+		opts...)
+	if err != nil {
+		zlog.Error(ctx).Err(err).Send()
+		return nil, err
+	}
+	indexReport := convertToIndexReport(clairReport)
+	indexReport.HashId = req.GetHashId()
+	// TODO Define behavior for indexReport.Err != "".
+	return indexReport, nil
+}
+
+// validateContainerImageRequest validates a container image request.
+func validateContainerImageRequest(req *v4.CreateIndexReportRequest) error {
+	if req == nil {
+		return errox.InvalidArgs.New("empty request")
+	}
+	if !strings.HasPrefix(req.GetHashId(), "/v4/containerimage/") {
+		return errox.InvalidArgs.Newf("invalid hash id: %q", req.GetHashId())
+	}
+	if req.GetContainerImage() == nil {
+		return errox.InvalidArgs.New("invalid resource locator for container image")
+	}
+	// Validate container image URL.
+	imgURL := req.GetContainerImage().GetUrl()
+	if imgURL == "" {
+		return errox.InvalidArgs.New("missing image URL")
+	}
+	u, err := url.Parse(imgURL)
+	if err != nil {
+		return errox.InvalidArgs.Newf("invalid image URL: %q", imgURL).CausedBy(err)
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
```

Comment:
Should we consider a default value or to explicitly require the client to tell us it is insecure? How often are image names prefixed with the scheme?

Reply:
This is based on the [decision made](https://github.com/stackrox/stackrox/blob/master/scanner/decisions/0001-scannerv4-apis.md) about the APIs.

Image names are not URLs, so they should not be prefixed with a locator scheme. Scanner accepts URLs, and the only allowed schemas for now are "http" and "https".

Reply:
When considering alternatives, there are two options to think about. The first option is to have the scanner only accept strict image references in the ContainerImageLocator instead of URLs. This would require additional fields, such as a bool for "secure", to convey the same information if other methods are added. The second option is to create different types of container image locators, like SecureContainerImageLocator, InsecureContainerImageLocator, and SomeOtherProcotolContainerImageLocator.

The current approach is based on the fact that URLs are more commonly used and easier to understand. However, if we choose to change the approach, it's important to update the records to explain why. Also, note that defaults and implicit values may not be as important since Scanner APIs are internal and not customer-facing.